src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs

```diff
@@ -0,0 +1,276 @@
+
+            static string GetLabelAndValue(string label, string value)
+            {
+                var safeKey = GetSafeLabelName(label);
+                var safeValue = GetSafeLabelValue(value);
+                return $"{safeKey}=\"{safeValue}\"";
+            }
+        }
+
+        private static string GetSafeName(string name, char[] firstCharNameCharset, char[] charNameCharset)
+        {
+            // https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+            //
+            // Metric names and labels
+            // Every time series is uniquely identified by its metric name and a set of key-value pairs, also known as labels.
+            // The metric name specifies the general feature of a system that is measured (e.g. http_requests_total - the total number of HTTP requests received). It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex [a-zA-Z_:][a-zA-Z0-9_:]*.
+            // Note: The colons are reserved for user defined recording rules. They should not be used by exporters or direct instrumentation.
+            // Labels enable Prometheus's dimensional data model: any given combination of labels for the same metric name identifies a particular dimensional instantiation of that metric (for example: all HTTP requests that used the method POST to the /api/tracks handler). The query language allows filtering and aggregation based on these dimensions. Changing any label value, including adding or removing a label, will create a new time series.
+            // Label names may contain ASCII letters, numbers, as well as underscores. They must match the regex [a-zA-Z_][a-zA-Z0-9_]*. Label names beginning with __ are reserved for internal use.
+            // Label values may contain any Unicode characters.
+
+            var sb = new StringBuilder();
+            var firstChar = name[0];
+
+            sb.Append(firstCharNameCharset.Contains(firstChar)
+                ? firstChar
+                : GetSafeChar(char.ToLowerInvariant(firstChar), firstCharNameCharset));
+
+            for (var i = 1; i < name.Length; ++i)
+            {
+                sb.Append(GetSafeChar(name[i], charNameCharset));
+            }
+
+            return sb.ToString();
+
+            static char GetSafeChar(char c, char[] charset) => charset.Contains(c) ? c : '_';
+        }
+
+        private static string GetSafeMetricName(string name) => GetSafeName(name, FirstCharacterNameCharset, NameCharset);
+
+        private static string GetSafeLabelName(string name) => GetSafeName(name, FirstCharacterLabelCharset, LabelCharset);
+
+        private static string GetSafeLabelValue(string value)
+        {
+            // label_value can be any sequence of UTF-8 characters, but the backslash
+            // (\), double-quote ("), and line feed (\n) characters have to be escaped
+            // as \\, \", and \n, respectively.
+
+            var result = value.Replace("\\", "\\\\");
```

Comment:
This would scan the string 3 times, with extra allocations. There might be an optimal way.